include/session/config/conversations.h

```diff
@@ -0,0 +1,222 @@
+/// setting fields in the convo, and then giving it to convos_set().
+bool convos_get_or_construct_legacy_closed(
+        const config_object* conf, convo_legacy_closed_group* convo, const char* id)
+        __attribute__((warn_unused_result));
+
+/// Adds or updates a conversation from the given convo info
+void convos_set_1to1(config_object* conf, const convo_one_to_one* convo);
+void convos_set_open(config_object* conf, const convo_open_group* convo);
+void convos_set_legacy_closed(config_object* conf, const convo_legacy_closed_group* convo);
+
+/// Erases a conversation from the conversation list.  Returns true if the conversation was found
+/// and removed, false if the conversation was not present.  You must not call this during
+/// iteration; see details below.
+bool convos_erase_1to1(config_object* conf, const char* session_id);
+bool convos_erase_open(
+        config_object* conf, const char* base_url, const char* room, unsigned const char* pubkey);
+bool convos_erase_legacy_closed(config_object* conf, const char* group_id);
+
+/// Returns the number of conversations.
+size_t convos_size(const config_object* conf);
+/// Returns the number of conversations of the specific type.
+size_t convos_size_1to1(const config_object* conf);
+size_t convos_size_open(const config_object* conf);
+size_t convos_size_legacy_closed(const config_object* conf);
+
+/// Functions for iterating through the entire conversation list.  Intended use is:
+///
+///     convo_one_to_one c1;
+///     convo_open_group c2;
+///     convo_legacy_closed_group c3;
+///     convos_iterator *it = convos_iterator_new(my_convos);
+///     for (; !convos_iterator_done(it); convos_iterator_advance(it)) {
+///         if (convos_it_is_1to1(it, &c1)) {
+///             // use c1.whatever
+///         } else if (convos_it_is_open(it, &c2)) {
+///             // use c2.whatever
+///         } else if (convos_it_is_legacy_closed(it, &c3)) {
+///             // use c3.whatever
+///         }
+///     }
+///     convos_iterator_free(it);
+///
+/// It is permitted to modify records (e.g. with a call to one of the `convos_set_*` functions) and
+/// add records while iterating.
+///
+/// If you need to remove while iterating then usage is slightly different: you must advance the
+/// iteration by calling either convos_iterator_advance if not deleting, or convos_iterator_erase to
+/// erase and advance.  Usage looks like this:
+///
+///     convo_one_to_one c1;
+///     convos_iterator *it = convos_iterator_new(my_convos);
+///     while (!convos_iterator_done(it)) {
+///         if (convo_it_is_1to1(it, &c1)) {
+///             bool should_delete = /* ... */;
+///             if (should_delete)
+///                 convos_iterator_erase(it);
+///             else
+///                 convos_iterator_advance(it);
+///         }
+///     }
+///     convos_iterator_free(it);
+///
+
+typedef struct convos_iterator {
```

Comment:
this iterator never is used as a non pointer so it can (and should) be a forward declared struct.

Reply:
Why?  Just to make it opaque?